### ISSUE TITLE
mAP 0 with compute_detection_metrics_main

### ISSUE THREAD

**Issue by JuliaChae** [waymo-research/waymo-open-dataset#193]
Hi there,

I have a question regarding offline 3D detection evaluation and the .bin file formatting. Following the provided Python template `create_prediction_file_example.py`, I generated prediction.bin as well as my own ground truth bin file as I just wanted to evaluate on a subset of the validation data that I had. When I ran `compute_detection_metrics_main` on the two files, I got mAP of 0 on all categories. To troubleshoot, I tried running `compute_detection_metrics_main` on the fake prediction and fake ground truth files provided in the main repo and got mAP of 0 again. When I evaluate the fake_ground_truth.bin on itself (`compute_detection_metrics_main ../fake_ground_truths.bin  ../fake_ground_truths.bin `, I once again get mAP 0 instead of expected 1. Following issue #121 , I downloaded the official ground truth .bin data for validation and ran it against itself, and that evaluates to 1 in all categories, as expected. 

I believe it must be a problem with prediction file generation, but I don't understand why evaluation on `fake_prediction.bin` and `fake_ground_truth.bin` doesn't work. In `create_prediction_file.py`, I am changing the following:
`
o.context_name = name
o.frame_timestamp_micros = timestamp

# Populating box and score.
box = label_pb2.Label.Box() 
box.center_x = gt_box[0]
box.center_y = gt_box[1]
box.center_z = gt_box[2]
box.length = gt_box[3]
box.width = gt_box[4]
box.height = gt_box[5]
box.heading = gt_box[6]
o.object.box.CopyFrom(box)
# This must be within [0.0, 1.0]. It is better to filter those boxes with
# small scores to speed up metrics computation.
o.score = 1
# Use correct type
class_name = 'TYPE_' + name
o.object.type = getattr(label_pb2.Label, class_name)
`

I'm wondering if the problem is due to not explicitly setting difficulty level when populating ground truth data but I thought the levels were automatically set if not specified. Please let me know how I can fix this problem. I'd be happy to provide any other information if there is confusion.

Many thanks in advance!
Julia

**Comment by xmyqsh** [comment]
```
if (o.object().num_lidar_points_in_box() <= 0) continue;
```
Be care of this line in `compute_detection_metrics_main.cc`.
`fake_ground_truth.bin` hasn't written the `num_lidar_points_in_box` field.

And, when you generate your own ground_truth.bin, you must write this field, to get a valid mAP result instead of zero.

Erasing `gt_objects_ori` to `gt_objects` conversion will also be OK, but you will get an inconsistent result will the Validation leaderboard.

Actually, there should be a demo code to generate `groud_truth.bin`, along with a warning on the `num_lidar_points_in_box` field to save our times.

Anyway, why not submit to the evaluation server directly? You can evaluate you validation result any times, with a little extra time of delay.

**Comment by JuliaChae** [comment]
Oh I see, thanks for bringing my attention to that. Is `num_lidar_points_in_box` field required for both prediction and ground truth files to get a valid mAP? 

Also I'll consider submitting to evaluation server when I'm using the entire validation dataset to evaluate. Right now I'm just using a subset so I needed to generate the gt_data accordingly. 


> ```
> if (o.object().num_lidar_points_in_box() <= 0) continue;
> ```
> 
> Be care of this line in `compute_detection_metrics_main.cc`.
> `fake_ground_truth.bin` hasn't written the `num_lidar_points_in_box` field.
> 
> And, when you generate your own ground_truth.bin, you must write this field, to get a valid mAP result instead of zero.
> 
> Erasing `gt_objects_ori` to `gt_objects` conversion will also be OK, but you will get an inconsistent result with the Validation leaderboard.
> 
> Actually, there should be a demo code to generate `groud_truth.bin`, along with a warning on the `num_lidar_points_in_box` field to save our times.
> 
> Anyway, why not submit to the evaluation server directly? You can evaluate you validation result any times, with a little extra time of delay.

**Comment by peisun1115** [comment]
num_lidar_points_in_box is only needed for gt.

**Comment by BeckerFelix** [comment]
I agree that it is confusing to be given an mAP of 0. The user is not aware of the `num_lidar_points_in_box` field that is missing in provided `fake_ground_truths.bin` file in https://github.com/waymo-research/waymo-open-dataset/tree/master/waymo_open_dataset/metrics/tools. 

This holds especially since https://github.com/waymo-research/waymo-open-dataset/blob/a476ab4b88d0c8cec5ef83b243caab2c80812cfe/waymo_open_dataset/metrics/tools/compute_tracking_metrics_main.cc#L38-L39

**Comment by peisun1115** [comment]
We are planning to update the data to pre-populate these information in our next version to make it easier for user along with some other usability improvements. Stay-tuned!

**Comment by JuliaChae** [comment]
Thank you for the help!
